Support Microsoft Entra ID with Redisson client library (#48)
Azure-Samples/azure-cache-redis-samples: quickstart/java-redisson-jcache/README.md

---
# Quickstart: Use Azure Cache for Redis in Java with Redisson Redis client and JCache

> In this quickstart, you incorporate Azure Cache for Redis into a Java app using the [Redisson](https://redisson.org/) Redis client and JCP stardand JCache API to have access to a secure, dedicated cache that is accessible from any application within Azure.

Reviewer: spelling: "stardand" should be "standard"?

> In this quickstart, you incorporate Azure Cache for Redis into a Java app using the [Redisson](https://redisson.org/) Redis client and JCP stardand JCache API to have access to a secure, dedicated cache that is accessible from any application within Azure.

Reviewer 2: ```suggestion
In this quickstart, you incorporate Azure Cache for Redis into a Java app using the [Redisson](https://redisson.org/) Redis client and JCP standard JCache API to have access to a secure, dedicated cache that is accessible from any application within Azure.
```

## Prerequisites

- Azure subscription - [create one for free](https://azure.microsoft.com/free/)
- [Apache Maven](https://maven.apache.org/download.cgi)
- [Use Microsoft Entra ID for cache authentication](https://learn.microsoft.com/azure/azure-cache-for-redis/cache-azure-active-directory-for-authentication)

> ## Set up the working environment

Reviewer: We should also link to docs that describe how to configure your environment such that DefaultAzureCredential will pick up the identity you want to use for the Redis connection

Reviewer 3: Addressed in 427529f .

Depending on your operating system, add environment variables for your cache's **Host name** and **USERNAME**. Open a command prompt, or a terminal window, and set up the following values:

```CMD
set REDIS_CACHE_HOSTNAME=<YOUR_HOST_NAME>.redis.cache.windows.net

> set USERNAME=<USERNAME>

Reviewer: We're entirely removing the option of using an access key? That raises the bar pretty high for a sample - now users will need to correctly configure Entra ID auth on their cache and client. Can we retain a code path for access keys too, or maybe leave some code comments indicating how to use them?

Author: Hello @philon-msft, thank you for your suggestion, I agree with you, and I will rework the samples today.

Author: Hi @philon-msft, I have updated the sample, I kept both auth with redis key and auth with Entra ID, thank you for your suggestion. Please tell me if you have any other concern.

---
# Quickstart: Use Azure Cache for Redis in Java with Redisson Redis client and JCache

In this quickstart, you incorporate Azure Cache for Redis into a Java app using the [Redisson](https://redisson.org/) Redis client and JCP standard JCache API to have access to a secure, dedicated cache that is accessible from any application within Azure.

## Prerequisites

- Azure subscription - [create one for free](https://azure.microsoft.com/free/)
- [Apache Maven](https://maven.apache.org/download.cgi)

> - [Use Microsoft Entra ID for cache authentication](https://learn.microsoft.com/azure/azure-cache-for-redis/cache-azure-active-directory-for-authentication)

Reviewer: Configuring Entra is not a prereq if you're using access key auth. This link should move to the "### Option02: Authentication with Microsoft Entra ID" section

Reviewer: This "[Use Microsoft Entra ID" line still appears in the Prerequisites section. Please move it down into the "Authentication with Microsoft Entra ID" section

## Set up the working environment

> ### Option01: Authentication with Redis Key

Reviewer: I'd use the term "Access Key" rather than "Redis Key"

Author: @philon-msft Thank you for reminding me that.

> ### Option01: Authentication with Redis Key

Reviewer: ```suggestion
### Option 1: Authentication with an Access Key
```

Depending on your operating system, add environment variables for your cache's **Host name** and **Primary access key**. Open a command prompt, or a terminal window, and set up the following values:

```CMD

> set REDIS_CACHE_HOSTNAME=<YOUR_HOST_NAME>.redis.cache.windows.net

Reviewer: Remove `.redis.cache.windows.net` everywhere - the host name already includes that

Reviewer 3: Fixed in 7d7c511.

Reviewer 3: Fixed in 7d7c511.

> set REDIS_CACHE_KEY=<YOUR_PRIMARY_ACCESS_KEY>

Reviewer: Recommend using the name `REDIS_ACCESS_KEY`

Author: Updated

> set AUTH_TYPE=RedisKey

Reviewer: Can we eliminate `AUTH_TYPE`, and instead just use access key auth if REDIS_ACCESS_KEY is defined, fall back to Entra auth if `REDIS_USER_NAME` is defined, and throw an error if neither is defined?

Author: Yes, thank you for your suggestion.

```bash
export REDIS_CACHE_HOSTNAME=<YOUR_HOST_NAME>.redis.cache.windows.net
export REDIS_ACCESS_KEY=<YOUR_PRIMARY_ACCESS_KEY>
```

Replace the placeholders with the following values:

- `<YOUR_HOST_NAME>`: The DNS host name, obtained from the *Properties* section of your Azure Cache for Redis resource in the Azure portal.

> - `<YOUR_PRIMARY_ACCESS_KEY>`: The primary access key, obtained from the *Access keys* section of your Azure Cache for Redis resource in the Azure portal.

Reviewer: There is no "*Access keys* section" - you need to get access keys from the Overview section

Reviewer 3: Fixed in 7d7c511.

- `<YOUR_HOST_NAME>`: The DNS host name, obtained from the *Properties* section of your Azure Cache for Redis resource in the Azure portal.
- `<YOUR_PRIMARY_ACCESS_KEY>`: The primary access key, obtained from the *Access keys* section of your Azure Cache for Redis resource in the Azure portal.

> ### Option02: Authentication with Microsoft Entra ID

Reviewer: ```suggestion
### Option 2: Authentication with Microsoft Entra ID
```

> Depending on your operating system, add environment variables for your cache's **Host name** and **USER_NAME**. Open a command prompt, or a terminal window, and set up the following values:

Reviewer: ```suggestion
Depending on your operating system, add environment variables for your cache's **Host name** and the **Username** of the Entra identity to sue for the Redis connection. Open a command prompt, or a terminal window, and set up the following values:
```

```CMD
set REDIS_CACHE_HOSTNAME=<YOUR_HOST_NAME>.redis.cache.windows.net
set USER_NAME=<USER_NAME>
set AUTH_TYPE=MicrosoftEntraID
```

```bash
export REDIS_CACHE_HOSTNAME=<YOUR_HOST_NAME>.redis.cache.windows.net

> export USER_NAME=<USER_NAME>

Reviewer: I'd recommend calling this `REDIS_USER_NAME` for clarity

Author: @philon-msft The reason why I used it is to make it consistent with [microsoft-entra-client-workflow](https://learn.microsoft.com/azure/azure-cache-for-redis/cache-azure-active-directory-for-authentication#microsoft-entra-client-workflow). 
I agreed that using REDIS_USER_NAME is more meaningful.

export AUTH_TYPE=MicrosoftEntraID
```

Replace the placeholders with the following values:

- `<YOUR_HOST_NAME>`: The DNS host name, obtained from the *Properties* section of your Azure Cache for Redis resource in the Azure portal.
- `<USER_NAME>`: Object ID of your managed identity or service principal.
  - You can get the **USER_NAME** by following 1-4 steps in the image: 

>     - ![user_name.png](./media/user_name.png)

Author: You're correct. Text instructions are often easier to maintain and update than screenshots.

```CMD
mvn compile exec:java -D exec.mainClass=example.demo.App
```

## Next Steps

> It is also possible to use `DefaultAzureCredential` to provide the identity to be used with the Redis connection. For more details, see [Azure authentication with Java and Azure Identity](https://learn.microsoft.com/en-us/azure/developer/java/sdk/identity).

Reviewer: This line should move into the "### Option02: Authentication with Microsoft Entra ID" section and say something like "Configure DefaultAzureCredential environment variables to use the identity that was granted access on your Azure Cache for Redis resource. For details see [Configure DefaultAzureCredential
](https://learn.microsoft.com/en-us/azure/developer/java/sdk/identity-azure-hosted-auth#configure-defaultazurecredential). For general information see [Azure authentication with Java and Azure Identity](https://learn.microsoft.com/en-us/azure/developer/java/sdk/identity)."
